pkg/controllers/keymanagementprovider_controller.go

```diff
@@ -91,17 +90,28 @@ func (r *KeyManagementProviderReconciler) Reconcile(ctx context.Context, req ctr
 		return ctrl.Result{}, err
 	}
 
+	// fetch certificates and store in map
 	certificates, certAttributes, err := provider.GetCertificates(ctx)
 	if err != nil {
 		writeKMProviderStatus(ctx, r, &keyManagementProvider, logger, isFetchSuccessful, err.Error(), lastFetchedTime, nil)
 		return ctrl.Result{}, fmt.Errorf("Error fetching certificates in KMProvider %v with %v provider, error: %w", resource, keyManagementProvider.Spec.Type, err)
 	}
 	keymanagementprovider.SetCertificatesInMap(resource, certificates)
+
+	// fetch keys and store in map
+	keys, keyAttributes, err := provider.GetKeys(ctx)
+	if err != nil {
```

Comment:
quick question: if we succeed to set certs but failed to get keys, then it fall into a state that partial data is set up. Should we roll back the certMap in this case? open for discussion

Reply:
great point. I agree with you. I will change it so that certs + keys only get added to map after all operations that can potentially fail finish.